### PR TITLE
Staged aggs 3

### PR DESCRIPTION
**Pull request by patrick-schultz**
Third piece of #6518; staged on #6532. Add extract aggregators pass using staged aggregators. Currently uses files for communicating between workers and master. This isn't ideal, but I wanted to try to get this merged and tested as quickly as possible before starting to modify anything.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -826,6 +826,13 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
   override def partitionCounts: Option[IndexedSeq[Long]] = child.partitionCounts
 
   protected[ir] override def execute(ctx: ExecuteContext): TableValue = {
+    try {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: execute once

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 2]
```diff
@@ -1,16 +1,217 @@
...
 
 import scala.language.{existentials, postfixOps}
 
 
 class UnsupportedExtraction(msg: String) extends Exception(msg)
 
+object TableMapIRNew {
+  def apply(tv: TableValue, newRow: IR): TableValue = {
+    val typ = tv.typ
+    val gType = typ.globalType
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: should be `tv.globals.t` I think

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: and then you don't need .physicalType below

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 3]
```diff
@@ -1,16 +1,217 @@
...
+import is.hail.rvd.{RVDContext, RVDType}
 import is.hail.utils._
 
 import scala.language.{existentials, postfixOps}
 
 
 class UnsupportedExtraction(msg: String) extends Exception(msg)
 
+object TableMapIRNew {
+  def apply(tv: TableValue, newRow: IR): TableValue = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: also needs `ctx: ExecuteContext` as an argument.

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 4]
```diff
@@ -1,16 +1,217 @@
...
+          rvb.start(gType.physicalType)
+          rvb.addAnnotation(gType, globalsBc.value)
+          rvb.end()
+        } else
+          0
+      val pathOff = region.appendString(initPath)
+      val f = initF(0)
+      Region.scoped { aggRegion =>
+        f.newAggState(aggRegion)
+        f(region, globals, false, pathOff, false)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: `f(region, tv.globals.value.offset, false, pathOff, false)`

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 5]
```diff
@@ -1,16 +1,217 @@
...
+    // 1. init op on all aggs and write out to initPath
+    val initPath = HailContext.get.getTemporaryFile(prefix = Some("scan-init"))
+
+    val (_, initF) = ir.CompileWithAggregators2[Long, Long, Unit](
+      extracted.aggs,
+      "global", gType.physicalType,
+      path.name, PString(),
+      Begin(FastIndexedSeq(extracted.init, extracted.writeSet(0, path, spec))))
+
+    Region.scoped { region =>
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: val region = ctx.region

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 6]
```diff
@@ -1,16 +1,217 @@
...
+      extracted.aggs,
+      "global", gType.physicalType,
+      "row", typ.rowType.physicalType,
+      extracted.seqPerElt)
+
+    val scanPartitionPaths = tv.rvd.collectPerPartition { (i, ctx, it) =>
+      val globalRegion = ctx.freshRegion
+      val globals =
+        if (scanSeqNeedsGlobals) {
+          val rvb = new RegionValueBuilder(globalRegion)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: `globalsBc.value.readRegionValue(ctx.freshRegion)`

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 7]
```diff
@@ -1,16 +1,217 @@
...
+    val (_, combOpF) = ir.CompileWithAggregators2[Long, Long, Unit](
+      extracted.aggs ++ extracted.aggs,
+      path.name, PString(), // next partition
+      path2.name, PString(), // result file
+      Begin(
+        extracted.readSet(1, path, spec) +:
+        Array.tabulate(nAggs)(i => CombOp2(i, nAggs + i, extracted.aggs(i))) :+
+        extracted.writeSet(0, path2, spec)))
+
+    val resultPaths = Region.scoped { region =>
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala]: use same region

**Comment by catoverdrive** [comment]
@patrick-schultz looks like some of the tests are failing because the implementation in Emit is missing in this PR. I don't know if you wanted to swap the order of the two PRs and write a manual test or two for the other one to test that the codegen is doing the correct thing?

**Comment by danking** [comment]
bump #6535 is in. Is this PR under active development? @catoverdrive @patrick-schultz

**Comment by catoverdrive** [comment]
Closing this since the changes went in as #6653. (Had to make some changes due to underlying changes in the aggregator stuff)

**Pull request closed without merging**